FindQmlLint: Prefer to use the Qt::qmllint target

ferdnyc opened the pull request:
If a target exists `Qt${QT_MAJOR_VERSION}::qmllint`, we can simply use that as our qmllint executable instead of searching around the path for one.

We have to keep the path-based discovery for Qt5, which doesn't have a target.

I also added a `qmllint_all`, to be the parent target of all of the linting targets. (Which are themselves backed by `add_custom_command(OUTPUT..` targets. It's targets all the way down.

Fixes #894

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KDAB/GammaRay?pullRequest=895) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KDAB/GammaRay?pullRequest=895) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KDAB/GammaRay?pullRequest=895) it.</sub>

ratijas reviewed (approved):
+1

works for me, thanks!

dantti commented:
I wonder if we should still try the old way, like, if no Qt6 qmllint is found it might still find the Qt5 one, maybe it's best to only rely on the TARGET one?

Waqar144 commented:
> maybe it's best to only rely on the TARGET one?

+1

I would assume the newer Qt versions set the properly.

winterz commented:
@ferdnyc I have no idea why this PR thinks you haven't agreed to the CLA.  If I look in cla-assistant I see you listed in there.

ferdnyc commented:
@dantti @Waqar144 

Honestly, my preference would be to dump the path logic and always use the target as well, yes. Say the word and I'll swap out this Find module in favor of a macro that just runs `Qt::qmllint`.

@winterz 

Yeah, it's weird. Not the first time. Let me try squashing pre-commit's contributions out of it.

ferdnyc commented:
Oh! I couldn't figure out what that `cmake-lint` error was about, because I thought it was cut off. But no, that's the entire message: "Missing COMMENT in statement which allows it". 

...That seems dumb. _Forcing_ a comment on every target, even simple order/aggregation ones, just bloats the output. But, OK, whatever...

ferdnyc commented:
@winterz And now it's happy. Though I can't say for certain it had anything to do with pre-commit, merely force-pushing the branch unchanged and getting it to re-scan might've worked just as well.

ferdnyc commented:
@dantti @Waqar144 @winterz 

Ahh, crap. We all assumed wrong. Even Qt 5.15.2 _doesn't_ have a `::qmllint` target. So we're back to still needing the path search, with Qt5. :disappointed:

ferdnyc commented:
(We can probably still dump the weird Debian logic, tho, considering the comment above it mentions qt4(!).)

dantti commented:
well CI with 5.15 was happy so I thought the TARGET option was working

ferdnyc commented:
> well CI with 5.15 was happy so I thought the TARGET option was working

Nah, I'd left the path discovery as a fallback, so it must've used it. That will always work correctly in CI, anyway, because the Right Version™ of `qmllint` is the only one installed. The issue @ratijas hit was the build discovering the wrong `qmllint` on a system with Qt5 and Qt6 installed side-by-side. (Like most any desktop Linux system.)

ferdnyc commented:
Well, since I took the time to clean up the logic for creating `qmllint` targets, when I thought I could just move it into `GammaRayMacrosInternal.cmake`, I'll migrate those enhancement back into the Find module and add them here.

Any objections if I also try jettisoning this weirdness?

```cmake
    if(QMLLINT_EXECUTABLE)
        if(NOT QMLLINT_IS_WORKING)
            # Try to fix common problems on Debian-based distros -- they provide /usr/bin/qmllint,
            # which is a symlink to /usr/lib/x86_64-linux-gnu/qt4/bin/qmllint (or the Qt5 version of it).
            # The actual executable is part of different package, so might not even be installed =>
            # double-check whether qmllint is working by executing it
            execute_process(
                COMMAND ${QMLLINT_EXECUTABLE} --version
                RESULT_VARIABLE _qmllint_result
                OUTPUT_QUIET ERROR_QUIET
            )
            if(_qmllint_result EQUAL 0)
                set(QMLLINT_IS_WORKING
                    TRUE
                    CACHE BOOL "Whether the found qmllint executable is actually usable" FORCE
                )
```

If `qmllint` isn't installed the `find_program()` should come up with nothing, and if it's the wrong version it's still gonna be able to successfully run `qmllint --version`. That check tells us nothing.

dantti commented:
From the comment on this check find_program does find qmllint, but it finds the link, which doesn't work beacuse it points nowhere, are you sure find_program() for a broken link still comes up with nothing?

I see no "FOLLOW_SYMLINKS" or similar at the docs so I guess it's still a valid check

ferdnyc commented:
> From the comment on this check find_program does find qmllint, but it finds the link, which doesn't work beacuse it points nowhere, are you sure find_program() for a broken link still comes up with nothing?

\*sigh\* I'd LIKE to say that I'm sure even Debian have updated their packaging since Qt4, and no longer install broken symlinks to uninstalled binaries -- but if they're still using qtchooser, I can't say that. So, back it goes.

ferdnyc commented:
Huh. Maybe it really IS pre-commit. I just force-pushed a nothing change that didn't remove pre-commit's changes from the branch. (I made an edit, made a second to revert it, and squashed it with my earlier commit.) ..And it still claims I haven't signed the CLA.

But earlier, when I squashed pre-commit's changes into my own and force-pushed, that immediately turned the CLA check green.

ferdnyc commented:
Well, I'm gonna experimentally squash out pre-commit's changes again, let's see what happens.

ferdnyc commented:
Instant green light on the CLA front. Interesting.

Something seems to be _up_ between the CLAssistant and pre-commit. They have _beef_.

winterz commented:
you could run `pre-commit install` in your gammaray checkout.  then you'll be prompted to fix (or the problems are sometimes auto-fixed for you) any problems beforehand.

dantti commented:
LGTM can we merge?

ferdnyc commented:
@winterz 
> you could run `pre-commit install` in your gammaray checkout. then you'll be prompted to fix (or the problems are sometimes auto-fixed for you) any problems beforehand.

More of a workaround than a fix, but good idea while CLAssistant and pre-commit are warring.

Tho i'm sort of tempted to _not_ — or at least, not quite yet — so I can black-box test a few more aspects of the problem.

(Like: If I push _my own_ commits to a branch **after** pre-commit, does that clear up the CLA issue? Is it an issue with the mere _presence_ of pre-commit commits anywhere on the branch? Or is it only a problem if the pre-commit commit is at the branch HEAD?)
